### ISSUE TITLE
Variable as offset parameter in Nx.take_diagonal/2 inside Nx.Defn

### ISSUE THREAD

**Issue by RD314** [elixir-nx/nx#836]
Hi, 

how to work with `offset` param as variable in `Nx.take_diagonal/2` inside `Nx.Defn`?

There is an sample code that illustrate the issue:
```
defn upper_triangle(n_diagonals, init_matrix ) do

  {matrix_new, _, _, _} =
    while {matrix = Nx.broadcast(Nx.tensor(0), Nx.shape(init_matrix)),
           init_matrix,
           curent_diagonal = 1,
           n_diagonals},
           curent_diagonal < n_diagonals + 1
    do
      upper_diag_mat =
        init_matrix
        |> Nx.take_diagonal(offset: curent_diagonal)
        |> Nx.make_diagonal(offset: curent_diagonal)

      {matrix + upper_diag_mat, init_matrix, curent_diagonal + 1, n_diagonals}

    end
    matrix_new
end
```
The code should create copy offset diagonals of the `init_matrix` from 1 up to `n_diagonals`.

The problem is that `curent_diagonal ` variable is a tensor and tensor as an offset is not processed properly in `Nx.take_diagonal(init_matrix, offset: curent_diagonal)`.

The same code as shown here for "defn" works well in "def" (after some rework around while).

How to pass the variable to offset parameter in Nx.Defn code in `while`?

**Comment by josevalim** [comment]
With the current implementation it is not possible, as we build the indices before hand. An implementation that does it dynamically would have to be written and contributed first. My suggestion is for you to write a `take_diagonal_offset` function, that behaves the way you want, and submit a PR. :)

**Comment by Benjamin-Philip** [comment]
I think the deeper question is access this issue posses is: how does someone access the value of a scalar tensors as an Integer/Float?

In my opinion, the correct way to solve this problem is to provide a **standard** way to access `current_diagonal` as an integer/float, which can then be passed into `take_diagonal`. 

Another thing that could be done, is add support for tensor offsets, but there are enough other functions in the Nx API that take integer or float as a parameter. It also doesn't feel like the right thing to do.

TL;DR

Implement `take_diagonal_offset`, but make sure it supports floats, and name it something more general.

**Comment by RD314** [comment]
There is `Nx.to_number/1` that can;t be used inside defn. It seemed to me as the solution. Maybe there is a deeper reason why there is no way how to convert tensor to number inside defn. I am curious...

**Comment by jonatanklosko** [comment]
> In my opinion, the correct way to solve this problem is to provide a standard way to access current_diagonal as an integer/float, which can then be passed into take_diagonal.

The issue is that in many cases (including this one), the runtime value impacts the tensors shape, and we need to know the shapes at compile time. In particular, in this example each iteration would have a different expression graph.

The implementation could be fixed by using a `transform` and `Enum.reduce` to effectively inline all the iterations, however that's definitely not the best idea.

The function you're looking for could be implemented like this:

```elixir
defn upper_triangle(matrix, opts \\ []) do
  opts = keyword!(opts, [:diagonals])
  diagonals = opts[:diagonals]

  n = Nx.axis_size(matrix, 0)

  triangular_mask =
    Nx.less_equal(
      Nx.iota(matrix, axis: 0),
      Nx.iota(matrix, axis: 1) - n + diagonals
    )

  Nx.select(triangular_mask, matrix, 0)
end
```

**Comment by jonatanklosko** [comment]
Actually, the number of diagonals can totally be a regular argument as well

```elixir
defn upper_triangle(matrix, diagonals) do
  n = Nx.axis_size(matrix, 0)

  triangular_mask =
    Nx.less_equal(
      Nx.iota(matrix, axis: 0),
      Nx.iota(matrix, axis: 1) - n + diagonals
    )

  Nx.select(triangular_mask, matrix, 0)
end
```
